## Importing Needed packages Loading Data

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split , cross_val_score


In [2]:
df=pd.read_csv('winequality-red.csv')
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

## Preparing Test and Training Data

In [3]:
x = df.iloc[:,:11]
y=df['quality']


In [4]:
x_train,x_test,y_train,y_test =train_test_split(x,y,test_size=0.1)

## Fitting The Model

In [5]:
model = RandomForestClassifier(n_estimators=100,max_depth=12,random_state=0, criterion = 'entropy')
model.fit(x_train,y_train)
y_preds=model.predict(x_test)
print(cross_val_score(model,x_train,y_train,cv=10).mean())

C:\Users\mamou\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:670: UserWarning: The least populated class in y has only 8 members, which is less than n_splits=10.
  warnings.warn(("The least populated class in y has only %d"


0.6789724164724165


In [6]:
(y_preds==y_test).mean()

0.725

In [7]:
(abs(y_preds-y_test)>1).mean()

0.01875

In [8]:
(abs(y_preds-y_test)==1).mean()

0.25625

## Model Analysis 
### The model predicted 72.5% accuarcy of 10 Wine qualities 
### The model predicted wine quality as either above graded or under graded by one grade 25.625% of the time 
### The model predicted wine quality as either above graded or under graded by two grades 1.875% of the time

## Other Approach 
### categorize Wine qualites for 0 , 1 as bad  2 , 3 as fair , 4 , 5 as good 6,7 as premium 8 , 9 as ideal and then assign them ordinal variable as 0 , 1 , 2 , 3 , 4

In [9]:
df['quality']=df['quality']//2

In [10]:
y=df['quality']

In [11]:
x_train,x_test,y_train,y_test =train_test_split(x,y,test_size=0.1)

In [12]:
new_model = RandomForestClassifier(n_estimators=100,max_depth=12,random_state=0, criterion = 'entropy')
model.fit(x_train,y_train)
y_preds=model.predict(x_test)
print(cross_val_score(model,x_train,y_train,cv=10).mean())

C:\Users\mamou\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:670: UserWarning: The least populated class in y has only 9 members, which is less than n_splits=10.
  warnings.warn(("The least populated class in y has only %d"


0.8005681818181818


In [13]:
(y_preds==y_test).mean()

0.8

## We see an expected improvement in the model as the number of classes is lower the margin of error is lower due to that more alike classes are joined together 
### lets try even more and classify bad and fair as bad , good and premium as good , and ideal as ideal and assign them the value of 0 , 1 , 2 

In [14]:
df['quality']=df['quality']//2

In [15]:
y=df['quality']

In [16]:
x_train,x_test,y_train,y_test =train_test_split(x,y,test_size=0.1)

In [17]:
new_model = RandomForestClassifier(n_estimators=100,max_depth=12,random_state=0, criterion = 'entropy')
model.fit(x_train,y_train)
y_preds=model.predict(x_test)
print(cross_val_score(model,x_train,y_train,cv=10).mean())

0.9826291763791764


In [18]:
(y_preds==y_test).mean()

0.99375

## As can be seen 99% acuarracy in classification due to the reduction of classes but that decreases the specifipty of wine quality